src/core/qgstextrenderer.cpp

```diff
@@ -3217,175 +3282,332 @@ void QgsTextRenderer::drawTextInternal( TextPart drawType,
+
+      Component subComponent;
+      subComponent.text = line;
+      subComponent.size = QSizeF( labelWidth, labelHeight );
+      subComponent.offset = QPointF( 0.0, -ascentOffset );
+      subComponent.rotation = -component.rotation * 180 / M_PI;
+      subComponent.rotationOffset = 0.0;
+
+      if ( drawType == QgsTextRenderer::Buffer )
+      {
+        QgsTextRenderer::drawBuffer( context, subComponent, format, fontMetrics );
+      }
+      else
+      {
+        // draw text, QPainterPath method
+        QPainterPath path;
+        path.setFillRule( Qt::WindingFill );
+        path.addText( 0, 0, format.scaledFont( context ), subComponent.text );
+
+        // store text's drawing in QPicture for drop shadow call
+        QPicture textPict;
+        QPainter textp;
+        textp.begin( &textPict );
+        textp.setPen( Qt::NoPen );
+        QColor textColor = format.color();
+        textColor.setAlphaF( format.opacity() );
+        textp.setBrush( textColor );
+        textp.drawPath( path );
+        // TODO: why are some font settings lost on drawPicture() when using drawText() inside QPicture?
+        //       e.g. some capitalization options, but not others
+        //textp.setFont( tmpLyr.textFont );
+        //textp.setPen( tmpLyr.textColor );
+        //textp.drawText( 0, 0, component.text() );
+        textp.end();
+
+        if ( format.shadow().enabled() && format.shadow().shadowPlacement() == QgsTextShadowSettings::ShadowText )
+        {
+          subComponent.picture = textPict;
+          subComponent.pictureBuffer = 0.0; // no pen width to deal with
+          subComponent.origin = QPointF( 0.0, 0.0 );
+
+          QgsTextRenderer::drawShadow( context, subComponent, format );
+        }
+
+        // paint the text
+        if ( context.useAdvancedEffects() )
+        {
+          context.painter()->setCompositionMode( format.blendMode() );
+        }
+
+        // scale for any print output or image saving @ specific dpi
+        context.painter()->scale( subComponent.dpiRatio, subComponent.dpiRatio );
+
+        switch ( context.textRenderFormat() )
+        {
+          case QgsRenderContext::TextFormatAlwaysOutlines:
+          {
+            // draw outlined text
+            _fixQPictureDPI( context.painter() );
+            context.painter()->drawPicture( 0, 0, textPict );
+            break;
+          }
+
+          case QgsRenderContext::TextFormatAlwaysText:
+          {
+            context.painter()->setFont( format.scaledFont( context ) );
+            QColor textColor = format.color();
+            textColor.setAlphaF( format.opacity() );
+            context.painter()->setPen( textColor );
+            context.painter()->setRenderHint( QPainter::TextAntialiasing );
+            context.painter()->drawText( 0, 0, subComponent.text );
+          }
+        }
+      }
+      context.painter()->restore();
+      i++;
     }
+  }
+  else if ( format.orientation() == QgsTextFormat::VerticalOrientation )
```

Comment:
Lots of duplicate code here. Is there anyway to reduce this?

Reply:
Yeah, I thought so too, but I can't really see how to reduplicate this. Open to ideas.

For the record: there will be a 3rd vertical orientation mode added in a follow up PR that'll set the vertical orientation based on the rotation. So horizontal, vertical, rotation-dependent / based on rotation (not settled on name yet).